src/test.cairo

```diff
@@ -0,0 +1,576 @@
+
+#[test]
+#[should_panic(expected: ('Account: invalid signature', 'ENTRYPOINT_FAILED'))]
+fn test_validate_declare_empty_signature() {
+    let account = setup_dispatcher(Option::Some(@SIGNED_TX_DATA()));
+    let empty_sig = array![];
+
+    testing::set_signature(empty_sig.span());
+
+    account.__validate_declare__(CLASS_HASH());
+}
+
+fn test_execute_with_version(version: Option<felt252>) {
+    let data = SIGNED_TX_DATA();
+    let account = setup_dispatcher(Option::Some(@data));
+    let erc20 = deploy_erc20(account.contract_address, 1000);
+    let recipient = contract_address_const::<0x123>();
+
+    // Craft call and add to calls array
+    let mut calldata = array![];
+    let amount: u256 = 200;
+    calldata.append_serde(recipient);
+    calldata.append_serde(amount);
+    let call = Call {
+        to: erc20.contract_address, selector: selectors::transfer, calldata: calldata.span()
+    };
+    let mut calls = array![];
+    calls.append(call);
+
+    // Handle version for test
+    if version.is_some() {
+        testing::set_version(version.unwrap());
+    }
+
+    // Execute
+    let ret = account.__execute__(calls);
+
+    // Assert that the transfer was successful
+    assert_eq!(erc20.balance_of(account.contract_address), 800, "Should have remainder");
+    assert_eq!(erc20.balance_of(recipient), amount, "Should have transferred");
+
+    // Test return value
+    let mut call_serialized_retval = *ret.at(0);
+    let call_retval = Serde::<bool>::deserialize(ref call_serialized_retval);
+    assert!(call_retval.unwrap());
+}
+
+#[test]
+fn test_execute() {
+    test_execute_with_version(Option::None(()));
+}
+
+#[test]
+fn test_execute_future_version() {
+    test_execute_with_version(Option::Some(MIN_TRANSACTION_VERSION + 1));
+}
+
+#[test]
+fn test_execute_query_version() {
+    test_execute_with_version(Option::Some(QUERY_VERSION));
+}
+
+#[test]
+#[should_panic(expected: ('Account: invalid tx version', 'ENTRYPOINT_FAILED'))]
+fn test_execute_invalid_query_version() {
+    test_execute_with_version(Option::Some(QUERY_OFFSET));
+}
+
+#[test]
+fn test_execute_future_query_version() {
+    test_execute_with_version(Option::Some(QUERY_VERSION + 1));
+}
+
+#[test]
+#[should_panic(expected: ('Account: invalid tx version', 'ENTRYPOINT_FAILED'))]
+fn test_execute_invalid_version() {
+    test_execute_with_version(Option::Some(MIN_TRANSACTION_VERSION - 1));
+}
+
+#[test]
+fn test_validate() {
+    let calls = array![];
+    let account = setup_dispatcher(Option::Some(@SIGNED_TX_DATA()));
+
+    let is_valid = account.__validate__(calls);
+    assert_eq!(is_valid, starknet::VALIDATED);
+}
+
+#[test]
+#[should_panic(expected: ('Account: invalid signature', 'ENTRYPOINT_FAILED'))]
+fn test_validate_invalid() {
+    let calls = array![];
+    let mut data = SIGNED_TX_DATA();
+    data.transaction_hash += 1;
+    let account = setup_dispatcher(Option::Some(@data));
+
+    account.__validate__(calls);
+}
+
+#[test]
+fn test_multicall() {
```

Comment:
Could you please also add a test checking the full flow? Making sure we get the correct response from the `storage.cairo` contract and that the balance are updated correctly if the validation passes?